typing/typedecl.ml

```diff
@@ -1629,63 +1630,147 @@ let check_unboxable env loc ty =
+        raise(Error(pprim_type.ptyp_loc, Null_arity_external));
       if !Clflags.native_code
       && prim.prim_arity > 5
       && prim.prim_native_name = ""
-      then raise(Error(valdecl.pval_type.ptyp_loc, Missing_native_external));
+      then raise(Error(pprim_type.ptyp_loc, Missing_native_external));
       check_unboxable env loc ty;
-      { val_type = ty; val_kind = Val_prim prim; Types.val_loc = loc;
-        val_attributes = valdecl.pval_attributes;
+      { val_type = ty; val_kind = Val_prim (prim, Safe); Types.val_loc = loc;
+        val_attributes = primdesc.pprim_attributes;
         val_uid = Uid.mk ~current_unit:(Env.get_current_unit ());
       }
-  in
-  let (id, newenv) =
-    Env.enter_value valdecl.pval_name.txt v env
-      ~check:(fun s -> Warnings.Unused_value_declaration s)
-  in
-  let desc =
-    {
-     val_id = id;
-     val_name = valdecl.pval_name;
-     val_desc = cty; val_val = v;
-     val_prim = valdecl.pval_prim;
-     val_loc = valdecl.pval_loc;
-     val_attributes = valdecl.pval_attributes;
-    }
-  in
-  desc, newenv
+    in
+    let (id, newenv) =
+      Env.enter_value primdesc.pprim_name.txt v env
+        ~check:(fun s -> Warnings.Unused_value_declaration s)
+    in
+    let desc =
+      {
+       val_id = id;
+       val_name = primdesc.pprim_name;
+       val_val = v;
+       val_ext = Tval_prim_decl (cty, pprim_prim);
+       val_loc = primdesc.pprim_loc;
+       val_attributes = primdesc.pprim_attributes;
+      }
+    in
+    desc, newenv
+  | Pprim_alias (pprim_type, pprim_ident) ->
+    let (_ : Path.t), v = Env.lookup_value ~use:true ~loc pprim_ident.txt env in
+    let raise_alias_error kind =
+      raise(Error(pprim_ident.loc,
+                  Primitive_alias_does_not_refer_to_primitive kind))
+    in
+    (match v.val_kind with
+     | Val_prim (desc, _) as val_kind ->
+       let cty, v =
+         match pprim_type with
+         | None -> None, v
+         | Some pprim_type ->
+           let cty = Typetexp.transl_type_scheme env pprim_type in
+           let raise_unification_error env err =
+             raise(Error(cty.ctyp_loc,
+                         Primitive_alias_unification (env, err)))
+           in
+           let val_kind =
+             match
+               Ctype.matches ~expand_error_trace:true env
+                 cty.ctyp_type v.val_type
+             with
+             | () -> val_kind
+             | exception (Ctype.Matches_failure _)
+               when Env.is_in_signature env ->
+               (match
+                  Ctype.matches_gadt ~expand_error_trace:true env
+                    cty.ctyp_type v.val_type
+                with
+                | () -> Val_prim (desc, Unsafe_in_recmod)
+                | exception (Ctype.Matches_failure (env, err)) ->
+                  raise_unification_error env err)
+             | exception (Ctype.Matches_failure (env, err)) ->
+               raise_unification_error env err
```

Comment:
The control flow here (and perhaps also the logic) is rather complicated.  If there's a way to simplify it, that'd be good, but I'd appreciate at least seeing a comment explaining the aim of using `Ctype.matches`, `Ctype.matches_gadt` in this combination.